AEAA-363: Vulnerability / Advisory / CVSS data model refactor + public CVSS 2.0/3.1/4.0 implementations

YanWittmann opened the pull request:
This pull request introduces a major refactor of our data models for vulnerabilities and security advisories, along with the public release of our CVSS implementations for versions 2.0, 3.1, and 4.0.

---

## CVSS Standard implementation

Previously, the CVSS Vector management system was closed-source; with this pull request, this will change. This system allows for tracking an arbitrary amount of CVSS vectors on vulnerabilities and security advisories.

This means, the CVSS vectors are not evaluated and merged (potentially conceptually incorrectly) in early stages of the internal vulnerability enrichment process, where not enough information is present to make an informed decision on how to handle such vectors.

The new system evaluates these CVSS vectors only when they are required to be displayed or a score is needed, and evaluates them based on a defined ruleset that can be configured and overwritten by a user. This also allows for displaying the exact origin (source) of the vectors, including links to their hosting entity sites.

In addition to that, the CVSS:4.0 standard has been implemented and is now fully supported in the same ways as the previous versions.

![image](https://github.com/org-metaeffekt/metaeffekt-core/assets/37689635/59e30ea4-1352-40c0-a9f1-e17bc366ec43)

![image](https://github.com/org-metaeffekt/metaeffekt-core/assets/37689635/f03c670b-bd3d-46c2-be20-fd568f6e2c2d)

## Exact tracking of matched vulnerable configurations

The new vulnerability system now fully tracks all possible matching sources that a vulnerability could potentially come from, in addition to the exact vulnerable configuration and artifact that lead to matching the vulnerability.

This makes it possible to display more advanced information in the Vulnerability Assessment Dashboard and the Inventory Report.

![image](https://github.com/org-metaeffekt/metaeffekt-core/assets/37689635/e3933c03-f8e0-4bd7-93a6-2c403dee0089)

![image](https://github.com/org-metaeffekt/metaeffekt-core/assets/37689635/4755182f-8768-4e14-b941-a04b490fe8d6)

![image](https://github.com/org-metaeffekt/metaeffekt-core/assets/37689635/d6bdc5bf-7eb1-4df2-853f-62d682664c71)

![image](https://github.com/org-metaeffekt/metaeffekt-core/assets/37689635/37fb1707-d70a-4695-979a-620311344c54)

![image](https://github.com/org-metaeffekt/metaeffekt-core/assets/37689635/9dd9ea9c-12ad-49ad-b85c-2a78db92cff9)

## Separation of Security Advisories from Vulnerabilities

The inventory now no longer saves security advisories in a JSON Array in the Vulnerabilities sheet, but rather in a completely separate sheet Security Advisories. This has multiple advantages:

- Better separation of concerns
- Reducing redundancy of storing the same security advisory potentially many times
- Security advisories are now human-readable

![image](https://github.com/org-metaeffekt/metaeffekt-core/assets/37689635/18fcbe8a-2780-4525-8433-79c715beb974)

## Central Security Policy

Another big change is the introduction of the `CentralSecurityPolicyConfiguration` class as used in the `securityPolicy` parameters. It groups several parameters into a single class. Any properties that are not specified in the XML/JSON configurations are set to their default values.

Please note that any filtering properties from the security configuration only apply at the last steps of the pipelines: the report and the dashboard; but in those, they are applied in the exact same ways.

You can either explicitly set the `securityPolicy` parameters in the report generation and inventory enrichment mojos, or use the `securityPolicyFile` parameter, which will overwrite the `securityPolicy` with the contents of the referenced JSON file.

An example:

```json
{
  "insignificantThreshold": 7.0,
  "includeScoreThreshold": -1.0,
  "includeVulnerabilitiesWithAdvisoryProviders": ["all"],
  "includeAdvisoryTypes": ["all"],
  "vulnerabilityStatusDisplayMapperName": "abstracted",
  "cvssSeverityRanges": "None:pastel-gray:0.0:0.0,Low:strong-yellow:0.1:3.9,Medium:strong-light-orange:4.0:6.9,High:strong-dark-orange:7.0:8.9,Critical:strong-red:9.0:10.0",
  "cvssVersionSelectionPolicy": ["LATEST"],
  "baseCvssSelector": {"stats":[],"rules":[{"method":"ALL","stats":[],"selector":[{"host":["NVD"],"issuerRole":["CNA"],"issuer":["NVD"]},{"host":["Microsoft Corporation"],"issuerRole":["*"],"issuer":["*"]},{"host":["NVD"],"issuerRole":["CNA"],"issuer":["Microsoft Corporation"]},{"host":["GitHub, Inc."],"issuerRole":["*"],"issuer":["*"]},{"host":["NVD"],"issuerRole":["CNA"],"issuer":["GitHub, Inc."]},{"host":["NVD"],"issuerRole":["*"],"issuer":["*"]},{"host":["CERT-SEI"],"issuerRole":["*"],"issuer":["*"]},{"host":["not:Assessment"],"issuerRole":["*"],"issuer":["*"]}],"vectorEval":[]}],"vectorEval":[]},
  "effectiveCvssSelector": {"stats":[{"comparator":"EQUAL","action":"RETURN_NULL","attribute":"assessment","value":0}],"rules":[{"method":"ALL","stats":[],"selector":[{"host":["NVD"],"issuerRole":["CNA"],"issuer":["NVD"]},{"host":["Microsoft Corporation"],"issuerRole":["*"],"issuer":["*"]},{"host":["NVD"],"issuerRole":["CNA"],"issuer":["Microsoft Corporation"]},{"host":["GitHub, Inc."],"issuerRole":["*"],"issuer":["*"]},{"host":["NVD"],"issuerRole":["CNA"],"issuer":["GitHub, Inc."]},{"host":["NVD"],"issuerRole":["*"],"issuer":["*"]},{"host":["CERT-SEI"],"issuerRole":["*"],"issuer":["*"]},{"host":["not:Assessment"],"issuerRole":["*"],"issuer":["*"]}],"vectorEval":[]},{"method":"ALL","stats":[{"provider":"PRESENCE","attribute":"assessment","setType":"ADD"}],"selector":[{"host":["Assessment"],"issuerRole":["*"],"issuer":["all"]}],"vectorEval":[]},{"method":"LOWER","stats":[{"provider":"PRESENCE","attribute":"assessment","setType":"ADD"}],"selector":[{"host":["Assessment"],"issuerRole":["*"],"issuer":["lower"]}],"vectorEval":[]},{"method":"HIGHER","stats":[{"provider":"PRESENCE","attribute":"assessment","setType":"ADD"}],"selector":[{"host":["Assessment"],"issuerRole":["*"],"issuer":["higher"]}],"vectorEval":[]}],"vectorEval":[{"and":["not:IS_BASE_FULLY_DEFINED"],"action":"RETURN_NULL"}]}
}
```

---

These, and many other changes, make handling vulnerabilities, security advisories and CVSS vectors/scores more consistent and more easily (or at all) verifiable. I'm exited to see these changes in action.

YanWittmann committed:
AEAA-390: Initial move of CVSS Classes into Core

Signed-off-by: ywittmann <yan.wittmann@metaeffekt.com>

YanWittmann committed:
Made tests public

Signed-off-by: ywittmann <yan.wittmann@metaeffekt.com>

YanWittmann committed:
AEAA-388: Inventory writers will now use "Advisories" instead of "Cert" sheet name. Renamed CertMetaData to AdvisoryMetaData.

Signed-off-by: ywittmann <yan.wittmann@metaeffekt.com>

YanWittmann committed:
Merge branch 'master' into AEAA-363-ae-security-cvss

# Conflicts:
#	libraries/pom.xml

YanWittmann committed:
Merge branch 'master' into AEAA-363-ae-security-cvss

YanWittmann committed:
AEAA-363:
- renamed Cert sheet to Security Advisories
- added AdvisoryMetaData attributes
- restructured CVSS class packages
- renamed class Cvss4_0 to Cvss4P0
- added CvssSource class
- rearranged inventory sheets (sub order is now: artifacts, vulnerabilities, advisories, info)
- added formatted logging methods for AbstractModelBase instances

Signed-off-by: ywittmann <yan.wittmann@metaeffekt.com>

YanWittmann committed:
AEAA-363:
- extracted CvssEntity instance data into JSON files to be parsed at runtime with several specific ones being made available via static variables in KnownCvssEntities
- added class CvssVectorSet for management of multiple vectors with different sources
- fixes for the formatted logging methods for AbstractModelData collections

Signed-off-by: ywittmann <yan.wittmann@metaeffekt.com>

YanWittmann committed:
AEAA-363:
- advisories must now store their cvss vectors in a single colum "CVSS Vectors" as a SourcedCvssVectorSet
- added generic conditions for checking contents of maps and lists
- renamed "hamming distance" to "severity distance" in CVSS 4 implementation based on specification changes
- added JSON Schema for CVSS entities
- added CVSS entities for advisory providers
- proper equality, hash code and cloning methods for CVSS management classes
- removed dependencies to other metaeffekt modules
- added test cases for new CVSS management classes

Signed-off-by: ywittmann <yan.wittmann@metaeffekt.com>

YanWittmann committed:
AEAA-363:
- added better CVSS 4.0 doc comment
- removed ConditionTree classes, replacing it with a simple JSON Array
- added system for selecting and overlaying CVSS Vectors by source

Signed-off-by: ywittmann <yan.wittmann@metaeffekt.com>

YanWittmann committed:
AEAA-363:
- introduced score baking method for cvss
- fixed JavaDoc for CVSS 4
- extended CVSS Selector capabilities

Signed-off-by: ywittmann <yan.wittmann@metaeffekt.com>

YanWittmann committed:
Renamed CVSS Entity name provider interface to EntityNameProvider

Signed-off-by: ywittmann <yan.wittmann@metaeffekt.com>

YanWittmann committed:
Restructured CVSS class and package architecture:
- merged Sourced Vector information into actual vectors
- calculation methods will now return Double.NaN if requirements for property are not met
- CvssSelectionResult now stores actual vectors, since their source information has been merged already
added missing log4j2.xml for logging configuration.

Signed-off-by: ywittmann <yan.wittmann@metaeffekt.com>

YanWittmann committed:
Merge branch 'master' into AEAA-363-ae-security-cvss

YanWittmann committed:
AEAA-363:
- Moved data model from artifact-analysis into inventory-processor
- Started to integrate new data model into vulnerability report
- Rewrote StatisticsOverviewTable

Signed-off-by: ywittmann <yan.wittmann@metaeffekt.com>

YanWittmann committed:
AEAA-413: applied new advisory/vulnerability/cvss system to all report templates.
Countless minor adjustments and refactors regarding the report generation.

Signed-off-by: ywittmann <yan.wittmann@metaeffekt.com>

YanWittmann committed:
- Replaced all configuration parameters for the report generation with the security policy.
- Removed the AdvisoryData class in favor of individual security advisory classes.
- Added a GHSA matching reason to track the exact vulnerable configuration that lead to matching the vulnerability.
- Security policies can now be constructed from JSON files.
- Fixed inconsistent widths and invalid content of overview tables in report.

Signed-off-by: ywittmann <yan.wittmann@metaeffekt.com>

YanWittmann committed:
- Removed generic CvssVector type from cvss classes, since the benefit of having the additional type information was insignificant for the work of writing <?> or the specific type every time and having multiple signatures for CvssVector<T> and T itself.
- Added fallback to CVSS selector templates that picks any other than assessment vectors.

Signed-off-by: ywittmann <yan.wittmann@metaeffekt.com>

YanWittmann committed:
- Added missing license header files.
- Added method explaining a CvssSelector in plain text using a markdown-like structure.

Signed-off-by: ywittmann <yan.wittmann@metaeffekt.com>

YanWittmann committed:
- renamed attribute "Referenced Content IDs" to "Referenced Content Ids".
- renamed CVSS score/vector attributes and adjusted templates to fit new naming convention.
- removed "de" templates, since updating them would require more work than rewriting them.
- general code cleanup.

Signed-off-by: ywittmann <yan.wittmann@metaeffekt.com>

YanWittmann committed:
- renamed "Referenced Content Ids" once more to only "Referenced Ids".
- changed the way the inventory report vulnerability overview table displays its advisories and matched vulnerable configurations.

Signed-off-by: ywittmann <yan.wittmann@metaeffekt.com>

YanWittmann committed:
Minor adjustments to labels in inventory report

Signed-off-by: ywittmann <yan.wittmann@metaeffekt.com>

YanWittmann committed:
Merge branch 'master' into AEAA-363-ae-security-cvss

# Conflicts:
#	libraries/ae-inventory-processor/src/main/resources/META-INF/templates/de/asset-report-bom/macros/tpc_inventory-license.dita.vm
#	libraries/ae-inventory-processor/src/main/resources/META-INF/templates/de/asset-report-bom/tpc_asset-licenses.dita.vt

YanWittmann committed:
- added NaN check to the CVSS score normalization to allow for passing-through unavailable scores more reliably.
- small json format change to windows extraction.

Signed-off-by: ywittmann <yan.wittmann@metaeffekt.com>

YanWittmann committed:
GitHub action build fix attempt

Signed-off-by: ywittmann <yan.wittmann@metaeffekt.com>

YanWittmann commented:
> GitHub action build fix attempt

Amazing, it worked

karsten-klein commented on libraries/ae-inventory-processor/src/main/java/org/metaeffekt/core/inventory/processor/reader/AbstractInventoryReader.java:
```diff
@@ -21,6 +21,7 @@
 import org.apache.poi.ss.usermodel.DataFormatter;
 import org.apache.poi.ss.usermodel.Row;
 import org.metaeffekt.core.inventory.processor.model.*;
+import org.metaeffekt.core.inventory.processor.report.model.aeaa.AeaaInventoryAttribute;
```
Is this import needed. Inventory should not require report.

YanWittmann replied:
This import is used for the `AeaaInventoryAttribute.VULNERABILITY_REFERENCED_CONTENT_IDS` enum value for mapping the old way of writing `Referenced Content IDs` to the new one, which is `Referenced Ids`.

The issue with this is that whenever we add an attribute to the corresponding ´AbstractModelBase.Attribute´ enums, they are automatically added to the inventory when writing them out. If you check ´AeaaInventoryAttribute´ or it's Artifact-Analysis counterpart `InventoryAttribute` out, you will see a very large amount of attributes that are used in certain cases, but not always, so adding them to every inventory that is written would lead to much dead space in there.

I would also prefer moving them into the proper abstract model base implementations.

karsten-klein commented on libraries/ae-inventory-processor/src/main/java/org/metaeffekt/core/inventory/processor/report/model/aeaa/AeaaAmbDataClass.java:
We should consider a better package name and class prefix

YanWittmann replied:
I agree. All the data classes copied over share these properties:

- reside in `org.metaeffekt.core.inventory.processor.report.model.aeaa`
- have the exact name as their Artifact Analysis counterpart with the prefix `Aeaa`
- have a doc comment header addition of `Mirrors structure of <code>com.metaeffekt.artifact.analysis.vulnerability.enrichment.AeaaInventoryAttribute</code> until separation of inventory report generation from ae core inventory processor.`

We could / should perform one of these changes:

- change this prefix to something more meaningful
- remove the prefix and replace the indication of a mirrored class by something else

karsten-klein commented on libraries/ae-inventory-processor/src/main/resources/META-INF/templates/en/inventory-report-vulnerability/tpc_inventory-cert.dita.vt:
security advisors inside text not generally upper case; ok in titles

YanWittmann replied:
Changed to lowercase in all text except titles.

karsten-klein commented on libraries/ae-inventory-processor/src/main/resources/META-INF/templates/en/inventory-statistics-vulnerability/tpc_inventory-vulnerability-statistics.dita.vt:
```diff
@@ -32,19 +37,22 @@
-        <p id="statistics-table-modified-$advisory">
-#statisticsOverviewTableModified("vulnerabilities_statistics_table_modified_$advisory.toLowerCase()", "Effective Vulnerability Statistics with $advisory Advisories$reportContext.inContextOf()", "$advisory", $report.getOverviewTablesVulnerabilityStatusMappingFunction())
+        <p id="statistics-table-modified-$advisoryIdentifier">
+#statisticsOverviewTable("vulnerabilities_statistics_table_modified_$advisoryIdentifier.toLowerCase()", "Context Vulnerability Statistics with $advisoryName Advisories$reportContext.inContextOf()", $vulnerabilities, $advisory, true)
```
Are we sure that we do not need to escape attributes directly embedded in the text.

YanWittmann replied:
Do you mean the `advisoryIdentifier` / `advisoryName` in the id and title?

YanWittmann replied:
All vulnerability and advisory names are now escaped when used in the id attribute or in title tags.

YanWittmann committed:
Merge branch 'master' into AEAA-363-ae-security-cvss

# Conflicts:
#	libraries/ae-inventory-processor/src/test/java/org/metaeffekt/core/inventory/processor/RepositoryReportTest.java
#	libraries/ae-inventory-processor/src/test/java/org/metaeffekt/core/inventory/processor/report/AdvisoryDataTest.java

karsten-klein commented on libraries/ae-security/src/test/resources/META-INF/LICENSE:
Why is this required here; licenses are managen on core level automatically. Need to understand.

karsten-klein commented on libraries/ae-security/src/main/resources/log4j2.xml:
log4j is not our default logging implementation. The configuration must not be provided in src/main/resources. Maximum in src/test/resources.

YanWittmann replied:
Moved to src/test/resources.

karsten-klein commented on libraries/ae-security/src/main/resources/META-INF/LICENSE:
License should be managed automatically. Need to understand.

YanWittmann replied:
The LICENSE file is automatically created when I build the module in those two directories. I did not add them in the main/resources/META-INF/LICENSE and test/resources/META-INF/LICENSE. It seems as though it is added by the `<artifactId>license-maven-plugin</artifactId>` plugin.

karsten-klein commented on libraries/ae-security/.gitignore:
A local .gitignore should not be required.

karsten-klein commented on libraries/ae-security/pom.xml:
```diff
@@ -0,0 +1,71 @@
+        <dependency>
+            <groupId>org.apache.logging.log4j</groupId>
+            <artifactId>log4j-core</artifactId>
+            <optional>true</optional>
```
Change scope to test for log4j dependencies

karsten-klein reviewed (commented):
Hi Yan,

we need to at least add javadoc on class level. Also the setup regarding the license header plugin, license handling and logging configuration must be revised.

YanWittmann committed:
- AEAA-424: ContentIdentifiers now accept both well-formed name and enum name when parsing from string. Configuration parameters therefore now accept all written forms.
- AEAA-423: Introduced an "includeAdvisoryProviders" parameter that, similar to the "includeAdvisoryTypes" parameter, that will filter the displayed advisories by their provider.
- Normalized "all" "ALL" "any" "ANY" handling for advisory types and providers to use a central method.
- Fixed inventory report vulnerability summary table having an incorrect amount of morerows.
- renamed "Security Advisories" in inventory report plain text to "security advisories".

Signed-off-by: ywittmann <yan.wittmann@metaeffekt.com>

YanWittmann committed:
AEAA-425: Escaping of vulnerability and advisory ids in the Inventory Report

Signed-off-by: ywittmann <yan.wittmann@metaeffekt.com>

YanWittmann committed:
- added JavaDoc comments on class level for new classes.
- removed/moved unnecessary/generated files.
- made junit and log4j implementations only scope test.

Signed-off-by: ywittmann <yan.wittmann@metaeffekt.com>

YanWittmann committed:
Added table summaries for javadoc so that build no longer fails

Signed-off-by: ywittmann <yan.wittmann@metaeffekt.com>

YanWittmann committed:
Wouldn't it be great if the build told you all the cases of invalid JavaDoc at once and not one at a time?
Added more missing caption tags.

Signed-off-by: ywittmann <yan.wittmann@metaeffekt.com>

YanWittmann committed:
- Improved CVSS implementation code.
- Fixed rounding issue in CVSS:4.0 implementation.
- Fixed incorrect component order in CVSS:4.0 toString().
- Added more test data for CVSS validation.

Signed-off-by: ywittmann <yan.wittmann@metaeffekt.com>

YanWittmann committed:
- Implemented several utility methods for CVSS classes.
- Made all CVSS vector attributes implement a CvssVectorAttribute interface for consistent handling.
- Added a UniversalCvssCalculatorLinkGenerator class that can generate links to instances of https://github.com/org-metaeffekt/metaeffekt-universal-cvss-calculator.

Signed-off-by: ywittmann <yan.wittmann@metaeffekt.com>

YanWittmann committed:
Merge branch 'master' into AEAA-363-ae-security-cvss

# Conflicts:
#	libraries/ae-inventory-processor/src/main/java/org/metaeffekt/core/inventory/processor/adapter/NpmPackageLockAdapter.java
#	libraries/ae-inventory-processor/src/main/java/org/metaeffekt/core/inventory/processor/model/AbstractModelBase.java
#	libraries/ae-inventory-processor/src/main/java/org/metaeffekt/core/inventory/processor/patterns/ComponentPatternProducer.java
#	libraries/ae-inventory-processor/src/main/java/org/metaeffekt/core/inventory/processor/writer/AbstractInventoryWriter.java

YanWittmann committed:
- Added utility methods to CVSS classes.
- UniversalCvssCalculatorLinkGenerator: Refactored method names, added a gzip compressed, base64 encoded url parameter variant that, in many cases, generates a shorter version of the link. At least version 1.0.5 of the web calculator is required for this.

Signed-off-by: ywittmann <yan.wittmann@metaeffekt.com>

YanWittmann committed:
Removed www prefix from cvss calculator link generator base url

Signed-off-by: ywittmann <yan.wittmann@metaeffekt.com>

YanWittmann committed:
- Added support for a "priority" value in the vulnerability status history entries that allows for controlling the placement of the status by the user.
- Adjusted UniversalCvssCalculatorLinkGenerator base url.

Signed-off-by: ywittmann <yan.wittmann@metaeffekt.com>

YanWittmann commented:
... it's been a long time coming!
